include/swift/Sema/ConstraintLocator.h

```diff
@@ -508,6 +508,26 @@ public: \
 /// A base class for custom path elements that store numeric values.
 template <unsigned NumValues>
 class StoredIntegerElement: public LocatorPathElt {
+  static constexpr unsigned valueWidth() {
+    switch (NumValues) {
+    default:
+      return 16;
+    case 1:
+      return 64;
+    case 2:
+      return 32;
+    }
+  }
```

Comment:
My main concern with this is that bumping the number of stored integers on a path element after the fact has the potential to truncate values we previously accepted. Maybe that's a worthwhile tradeoff, but I think there should at least be a doc comment on `StoredIntegerElement` to warn of this

Reply:
We are not actually bumping it though, the problem was that in case of two values the first value was only 16 bits, which didn’t make much sense…

Reply:
Right, I mean if you have a `StoredIntegerElement<1>` today, you can store values up to 64 bits, but if in the future you need to bump it to a `StoredIntegerElement<2>`, you're now potentially truncating values you previously allowed

Reply:
That was the case previously, always need to be careful with that, luckily locators don’t frequently change that way…

Reply:
Hmm was it? It was my understanding that was always stored 16 bit values no matter how many

Reply:
Yeah, if you go from 1 to 2, that would shift your first value by 16, and 32 bits at 3, but I think it makes more sense to get 32/32 bits at 2

Reply:
Right but the width of the value always would have been 16 right? Or am I totally misunderstanding?

Reply:
The first value would be left shifted by 16 which means that we'd loose first 16 bits, so the first value would be 48 bits (assuming `unsigned long`) and the second one would be 16 bits stored at the tail.

Reply:
What this PR does is truncate first 32 bits and store two 32-bit values for `<2>` case.

Reply:
> The first value would be left shifted by 16 which means that we'd loose first 16 bits, so the first value would be 48 bits (assuming unsigned long) and the second one would be 16 bits stored at the tail.

IIRC we always masked to 16 bits

Reply:
The case in https://github.com/apple/swift/issues/61723 was even more fun. We stored number of elements in the tuple in 16 bits. And it was `StoredIntegerElement<1>` there.

So, after the patch we can store 1 64-bit value or 2 32-bit values or 3 16-bit values. In the reality since the API receives plain `unsigned` we will be limited to 32 bits.

Reply:
> which might not be immediately obvious to the person bumping it

I agree, we need to be vigilant about that and maybe change constructors to accept appropriately sized arguments to the data what is going to be stored, it might have it a bit inconvenient to use but at least it would be safer...

Reply:
We can change ctors, yes. E.g.:

```
  StoredIntegerElement(ConstraintLocator::PathElementKind kind, uint32_t value0, uint32_t value1)
    : LocatorPathElt(kind, (value0 << 16 | value1)) {
    : LocatorPathElt(kind, packValue<0>(value0) | packValue<1>(value1)) {
    assert(value0 == getValue<0>() && "value0 truncated");
    assert(value1 == getValue<1>() && "value1 truncated");
  }

  StoredIntegerElement(ConstraintLocator::PathElementKind kind, uint16_t value0,
                       uint16_t value1, uint16_t value2)
```

Reply:
Right, we won't need the assertions, I think that's a good thing, the reason why parameters are `unsigned` - arguments are usually unsigned as well, so we'd have to change the call sites which could be annoying.

Reply:
@asl If you would like to work on that, I'd be happy to review.

Reply:
already started